src/leaderboard/leaderboard_resource.rs

```diff
@@ -0,0 +1,128 @@
+
+use super::messages::data::IntoIds;
+
+/// Represents a remote leaderboard resource.
+pub struct Leaderboard {
+    data_clients: Vec<SLbClient<InterceptedService<Channel, HeaderInterceptor>>>,
+    deadline: Duration,
+    cache_name: String,
+    leaderboard_name: String,
+}
+
+impl Leaderboard {
+    /// Delete a leaderboard.
+    pub async fn delete(&self) -> MomentoResult<DeleteResponse> {
+        let request = DeleteRequest::new();
+        request.send(self).await
+    }
+
+    /// Fetch elements from a leaderboard by rank.
+    pub async fn fetch_by_rank(
+        &self,
+        rank_range: impl Into<RankRange>,
+    ) -> MomentoResult<FetchResponse> {
+        let request = FetchByRankRequest::new(rank_range);
+        request.send(self).await
+    }
+
+    /// Get elements from a leaderboard by score.
+    pub async fn fetch_by_score(
+        &self,
+        score_range: impl Into<ScoreRange>,
+    ) -> MomentoResult<FetchResponse> {
+        let request = FetchByScoreRequest::new(score_range);
+        request.send(self).await
+    }
+
+    /// Get the length of a leaderboard.
+    pub async fn length(&self) -> MomentoResult<LengthResponse> {
+        let request = LengthRequest::new();
+        request.send(self).await
+    }
+
+    /// Get rank of elements from a leaderboard using their element ids.
+    pub async fn get_rank<T: IntoIds>(&self, ids: T) -> MomentoResult<GetRankResponse> {
+        let request = GetRankRequest::new(ids);
+        request.send(self).await
+    }
+
+    /// Remove elements from a leaderboard using their element ids.
+    pub async fn remove_elements<T: IntoIds>(
+        &self,
+        ids: T,
+    ) -> MomentoResult<RemoveElementsResponse> {
+        let request = RemoveElementsRequest::new(ids);
+        request.send(self).await
+    }
+
+    /// Upsert (update/insert) elements into a leaderboard.
+    pub async fn upsert<E: IntoElements>(&self, elements: E) -> MomentoResult<UpsertResponse> {
+        let request = UpsertRequest::new(elements);
+        request.send(self).await
+    }
+
+    /* helper fns */
+    pub(crate) fn new(
+        data_clients: Vec<SLbClient<InterceptedService<Channel, HeaderInterceptor>>>,
+        deadline: Duration,
+        cache_name: impl Into<String>,
+        leaderboard_name: impl Into<String>,
+    ) -> Self {
+        Self {
+            data_clients,
+            deadline,
+            cache_name: cache_name.into(),
+            leaderboard_name: leaderboard_name.into(),
+        }
+    }
+
+    pub(crate) fn next_data_client(
+        &self,
+    ) -> SLbClient<InterceptedService<Channel, HeaderInterceptor>> {
+        let next_index =
+            NEXT_DATA_CLIENT_INDEX.fetch_add(1, Ordering::Relaxed) % self.data_clients.len();
+        self.data_clients[next_index].clone()
+    }
+
+    pub(crate) fn deadline(&self) -> Duration {
+        self.deadline
+    }
+
+    pub(crate) fn cache_name(&self) -> &String {
```

Comment:
`&String` is not normal. Unless there's a specific reason you need a String reference you should return `&str` instead (here and elsewhere). Using &String makes it misleading.

Reply:
Fixed in [f31495c](https://github.com/momentohq/client-sdk-rust/pull/438/commits/f31495c5d101f527feba5c41137db0c9c0b1e959)